### PR TITLE
Fix S1751 FP: Triggered when returning a property inside try catch block

### LINKED ISSUE #2522: Fix S1751 FP: triggered when returning a property
### Description

returning a property in a try catch confuses the analyser

### Repro steps

```csharp
class C
{
	int P => throw new Exception();
	int M()
	{
		while (true)
		{
			try
			{
				return P;
			}
			catch (Exception ex)
			{
			}
		}
	}
}
```

### Expected behavior

No Exception

### Actual behavior

S1751: Refactor the containing loop to do more than one iteration

### Related information

* SonarC# Version: 7.14.0.8411
* Visual Studio Version: 16.2

### PR DESCRIPTION
**Pull request by csaba-sagi-sonarsource**
Fixes #2522

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnconditionalJumpStatement.vb [hunk 1]
```diff
@@ -380,5 +380,39 @@ Namespace Tests.Diagnostics
             Throw New NotImplementedException()
         End Sub
     End Class
+
+    Public Class ReproducerClass
+        ReadOnly Property FirstProperty As ReproducerClass
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnconditionalJumpStatement.vb:385]: _Nitpick:_ There should be EOL around `Class` and `End Class`
```suggestion
    Public Class ReproducerClass
    
        ReadOnly Property FirstProperty As ReproducerClass
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnconditionalJumpStatement.vb [hunk 2]
```diff
@@ -380,5 +380,39 @@ Namespace Tests.Diagnostics
...
+        Public Function MethodWithMemberAccess() As Int32
+              While True
+                Try
+                    Return FirstProperty.SecondProperty ' Compliant
+                Catch E As Exception
+
+                End Try
+            End While
+        End Function
+    End Class
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnconditionalJumpStatement.vb:416]: _Nitpick:_ NL before and after `End Class`
```suggestion
        End Function
        
    End Class
    
```

File: analyzers/src/SonarAnalyzer.Common/Rules/UnconditionalJumpStatementBase.cs [hunk 3]
```diff
@@ -19,29 +19,36 @@
...
     {
         protected const string DiagnosticId = "S1751";
-        protected const string MessageFormat = "Refactor the containing loop to do more than one iteration.";
+        private const string MessageFormat = "Refactor the containing loop to do more than one iteration.";
 
-        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+        private readonly DiagnosticDescriptor rule;
 
-        protected abstract LoopWalkerBase<TStatementSyntax, TLanguageKindEnum> GetWalker(SyntaxNodeAnalysisContext context);
+        protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UnconditionalJumpStatementBase.cs:39]: This and the next abstract property `LoopStatements` should be together. And they won't need empty line between them.

File: analyzers/src/SonarAnalyzer.Common/Rules/UnconditionalJumpStatementBase.cs [hunk 4]
```diff
@@ -99,16 +110,16 @@ public List<TStatementSyntax> GetRuleViolations()
         {
             var ancestors = node
                 .Ancestors()
-                .TakeWhile(n => !this.rootExpression.Equals(n))
+                .TakeWhile(n => !rootExpression.Equals(n))
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UnconditionalJumpStatementBase.cs:113]: Line 108/109: Empty line in the middle of method signature makes it hard to read

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 5]
```diff
@@ -75,15 +70,43 @@ private class LoopWalker : LoopWalkerBase<StatementSyntax, SyntaxKind>
...
 
-            public override void Visit()
+            protected override bool IsAccessToClassMember(StatementSyntax node)
             {
-                var csWalker = new CsLoopwalker(this);
-                csWalker.SafeVisit(this.rootExpression);
+                var returnStatementExpression = ((ReturnStatementSyntax)node).Expression;
+                if (returnStatementExpression is IdentifierNameSyntax identifier
+                    && semanticModel.GetSymbolInfo(identifier) is { } symbolInfo
+                    && symbolInfo.Symbol is { } symbol
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs:85]: `SymbolInfo` is a struct so it cannot be `null`. You can go directly for it's symbol
```suggestion
                    && semanticModel.GetSymbolInfo(identifier).Symbol is { } symbol
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 6]
```diff
@@ -75,15 +70,43 @@ private class LoopWalker : LoopWalkerBase<StatementSyntax, SyntaxKind>
...
+                    && symbol.Kind == SymbolKind.Property)
+                {
+                    return true;
+                }
+
+                // We are checking for memberAccessExpression to catch NullReferenceException.
+                return returnStatementExpression is MemberAccessExpressionSyntax memberAccessExpression;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs:92]: Unused variable
```suggestion
                return returnStatementExpression is MemberAccessExpressionSyntax;
```

Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs:92]: We should also check for `ElementAccessExpression` for `return arr[0]`

Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs:92]: We should check `returnStatementExpression.DescendantNodesAndSelf().Any(x => x.IsAnyKind(...` to cover all nested cases. Like `(((arr[0])))` or other unexpected syntaxes like `return x = member.Item`

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Improvements suggested

**Comment by csaba-sagi-sonarsource** [comment]
WDYT about introducing ISyntaxHelper to the LanguageFacade to remove the duplication?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnconditionalJumpStatement.vb (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnconditionalJumpStatement.vb]: I wrote 
>and after `End Class`
🙂

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 7]
```diff
@@ -75,15 +74,34 @@ private class LoopWalker : LoopWalkerBase<StatementSyntax, SyntaxKind>
...
             {
+                var csWalker = new CsLoopwalker(this);
+                csWalker.SafeVisit(rootExpression);
             }
 
-            public override void Visit()
+            protected override bool IsPropertyAccess(StatementSyntax node) =>
+                node.DescendantNodes().Any(x => x.IsKind(SyntaxKind.IdentifierName)
+                                                && x is IdentifierNameSyntax identifier
+                                                && semanticModel.GetSymbolInfo(identifier).Symbol is { } symbol
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs:87]: We don't need `IsKind` check before `is IdentifierNameSyntax` in this case. That can simplify the syntax like this:
```suggestion
                node.DescendantNodes().OfType<IdentifierNameSyntax>().Any(x => semanticModel.GetSymbolInfo(x).Symbol is { } symbol
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 8]
```diff
@@ -75,15 +74,34 @@ private class LoopWalker : LoopWalkerBase<StatementSyntax, SyntaxKind>
...
+                                                && symbol.Kind == SymbolKind.Property);
+
+            protected override bool IsAnyKind(SyntaxNode node, ISet<SyntaxKind> syntaxKinds) => node.IsAnyKind(syntaxKinds);
+            protected override bool IsReturnStatement(SyntaxNode node) => node.IsKind(SyntaxKind.ReturnStatement);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs:91]: Formatting (EOL between them, as they are not properties and they are not trivial)
```suggestion
            protected override bool IsAnyKind(SyntaxNode node, ISet<SyntaxKind> syntaxKinds) => 
                node.IsAnyKind(syntaxKinds);
            
            protected override bool IsReturnStatement(SyntaxNode node) => 
                node.IsKind(SyntaxKind.ReturnStatement);
```

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/UnconditionalJumpStatement.cs [hunk 9]
```diff
@@ -83,15 +79,36 @@ private class LoopWalker : LoopWalkerBase<StatementSyntax, SyntaxKind>
...
 
             public override void Visit()
             {
                 var vbWalker = new VbLoopwalker(this);
-                vbWalker.SafeVisit(this.rootExpression);
+                vbWalker.SafeVisit(rootExpression);
+            }
+
+            protected override bool IsPropertyAccess(StatementSyntax node) =>
+                node.DescendantNodes().Any(x => x.IsKind(SyntaxKind.IdentifierName)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Rules/UnconditionalJumpStatement.cs:91]: Same as C# in this file

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Last round

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**